packages/react/src/components/TextArea/TextArea.tsx

```diff
@@ -155,22 +161,67 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
   const prefix = usePrefix();
   const { isFluid } = useContext(FormContext);
   const { defaultValue, value, disabled } = other;
+
   const [textCount, setTextCount] = useState(
-    defaultValue?.toString().length || value?.toString().length || 0
+    defaultValue
+      ? counterMode === 'character'
+        ? defaultValue.toString().length
+        : defaultValue.toString().match(/\w+/g)?.length || 0
+      : value
+      ? counterMode === 'character'
+        ? value.toString().length
+        : value.toString().match(/\w+/g)?.length || 0
+      : 0
   );
   const { current: textAreaInstanceId } = useRef(getInstanceId());
 
   const textareaProps: {
     id: TextAreaProps['id'];
+    onKeyDown: (evt: React.KeyboardEvent) => void;
     onChange: TextAreaProps['onChange'];
     onClick: TextAreaProps['onClick'];
     maxLength?: number;
   } = {
     id,
+    onKeyDown: (evt) => {
+      if (!other.disabled && counterMode === 'word') {
+        const key = evt.which;
+        if (textCount == maxCount && key == 32) {
+          evt.preventDefault();
+        }
+      }
+    },
     onChange: (evt) => {
-      if (!other.disabled && onChange) {
-        setTextCount(evt.target.value?.length);
-        onChange(evt);
+      if (!other.disabled) {
+        if (counterMode == 'character') {
+          setTextCount(evt.target.value.length);
+        } else if (counterMode == 'word') {
+          if (!evt.target.value) {
+            console.log('reset');
+            setTextCount(0);
+            return;
+          }
+
+          if (typeof maxCount !== 'undefined' && textareaRef.current !== null) {
+            const matchedWords = evt.target.value.match(/\w+/g);
+            if (matchedWords && matchedWords.length <= maxCount) {
+              textareaRef.current.removeAttribute('maxLength');
+              setTextCount(matchedWords.length);
+            } else if (matchedWords && matchedWords.length > maxCount) {
+              const first_max = evt.target.value
+                .split(/\s+/)
+                .slice(0, maxCount)
+                .join(' ');
+
+              setTextCount(maxCount);
+              textareaRef.current.value = first_max;
+            }
+            // todo: add handling if no matched words were found
```

Comment:
is a handler still necessary or can we remove this comment?

Reply:
done